Lib/shutil.py

```diff
@@ -550,7 +557,12 @@ def move(src, dst, copy_function=copy2):
             os.rename(src, dst)
             return
 
-        real_dst = os.path.join(dst, _basename(src))
+        # Using _basename instead of os.path.basename is important, as we must
+        # ignore any trailing slash to avoid the basename returning ''
+        # Forcing src to a string allows flexibility of objects being passed in,
+        # including Path objects
+        real_dst = os.path.join(dst, _basename(str(src)))
```

Comment:
Is there a reason to not use `os.fspath(src)` here?

Reply:
Do you mean to use `os.fspath(src)` instead of casting to a string?

I suppose it would guard users against passing in anything that isn't a string, bytes or path. Otherwise, if it is a path, `__fspath__` simply calls `str()` anyway. However, `_basename` relies on having a string as its argument, bytes would indeed break that.

Reply:
fspath() is better. I'm surprised there are no other functions in shutil.py with the same need.

Reply:
`fspath()` is used elsewhere in shutil.py, but it seems that only `shutil.move` needs to handle its basename differently. If we really want to use `fspath()` now, I can look at refactoring `_basename()`, otherwise I can add this to my list of things to improve for shutil.py.

Reply:
shutil doesnt explicitly document that operations don't support all path-like objects (`str`, `bytes`, and `os.PathLike`), though the `copy*` functions do say their arguments should be strings . For the most part shutil does work with path-like objects, but `_basename` doesn't. 

Since `_basename` breaks the abstraction of `os.path.basename`, it's `_basename` that needs to be fixed, not everywhere that it gets called, even if for now that's only in one place. It should use `path = os.fspath(path)`. If this returns `bytes`, it should also use `sep = os.fsencode(sep)`.

Reply:
So that's actually a typo in the PR name (it wouldn't let me edit it after I realized). This modifies `shutil.move` not `shutil.copy`, which doesn't specify that its arguments are strings. 

I'm not sure I'd say that `_basename` breaks the abstraction of `os.path.basename`, it just gives us part of the path with different requirements. But, I didn't write the original code :)

I'm happy to take another look at this though.